File: CarePackage/Software/PowerToys.cs
namespace CarePackage.Software;

public class PowerToys : BaseSoftware
{
    public override string           Key            { get; set; } = "microsoft-powertoys";
    public override string           Name           { get; set; } = "PowerToys";
    public override SoftwareCategory Category       { get; set; } = SoftwareCategory.Utility;
    public override string           DownloadName   { get; set; } = "PowerToysUserSetup-x64.exe";
    public override bool             IsArchive      { get; set; } = false;
    public override bool             ShouldCacheUrl { get; set; } = true;
    public override bool             RequiresAdmin  { get; set; } = false;
    public override Bitmap           Icon           { get; set; } = Resources.Icons.powertoys;
    public override string           Homepage       { get; set; } = "https://learn.microsoft.com/en-us/windows/powertoys";

    private readonly GitHubService _github;
    
    public PowerToys(IServiceProvider services)
    {
        _github = services.GetRequiredService<GitHubService>();
    }
    
    public override async Task<string> GetDownloadUrlAsync(CancellationToken ct)
    {
        var assets = await _github.GetLatestRepositoryReleaseAssetsAsync("microsoft", "PowerToys", ct);
        var asset  = assets.FirstOrDefault(a => a.Contains("PowerToysSetup-") && a.EndsWith("-x64.exe"));
        
        DownloadUrlResolveException.ThrowIf(asset is null);
        
        return asset;
    }
}

File: src/lib/software/powertoys.py
from PySide6.QtCore import Slot
from src.lib.software import BaseSoftware, SoftwareCategory
from src.lib.github_release_scraper import GithubReleaseScraper

class PowerToys(BaseSoftware):
    def __init__(self):
        super().__init__()

        self._gh = GithubReleaseScraper('microsoft', 'PowerToys', self)
        self._gh.releases_scraped.connect(self._on_releases_scraped)

        self.key = 'microsoft-powertoys'
        self.name = 'PowerToys'
        self.category = SoftwareCategory.Utility
        self.download_name = 'PowerToysUserSetup-x64.exe'
        self.is_archive = False
        self.should_cache_url = True
        self.requires_admin = False
        self.icon = 'powertoys.png'
        self.homepage = 'https://learn.microsoft.com/en-us/windows/powertoys'

    @Slot(list)
    def _on_releases_scraped(self, releases: list[str]):
        asset = next((release for release in releases if 'PowerToysSetup-' in release and release.endswith('-x64.exe')), None)
        if asset:
            self.url_resolved.emit(asset)
        else:
            self.url_resolve_error.emit(self.ResolveError.GitHubAssetNotFoundError)

    def resolve_download_url(self):
        self._gh.get_repo_releases()
